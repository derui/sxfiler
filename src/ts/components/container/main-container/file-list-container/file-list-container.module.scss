@import "../../../ui/commom-style/_colors.scss";
@import "../../../ui/commom-style/_variables.scss";
@import "../../../ui/commom-style/reset.module.css";

.root {
  display: grid;
  grid-template-rows: 100%;
  grid-template-columns: 50% auto 50%;

  color: $sol_base1;
  overflow: hidden;
  background-color: $sol_base03;

  font-size: 1rem;
}

.separator {
  width: 2px;
  background-color: $sol_base00;
  margin: 0 $nanoSpace;
}
