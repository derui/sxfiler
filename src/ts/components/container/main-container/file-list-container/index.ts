export * from "./file-list-container";
