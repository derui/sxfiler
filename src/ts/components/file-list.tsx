import * as React from "react";
import classNames from "classnames";

import * as Domain from "../domain/filer";
import FileItem from "./file-item";


interface HeaderProp {
  directory: string;
  focused: boolean;
}

/**
 * component definition for header of file list
 */
const Header : React.FunctionComponent<HeaderProp> = ({directory, focused}) => {
  const className = classNames("fp-FileList_Header", {
    "fp-FileList_Header-focused": focused,
  });

  return (<header className={className}>{directory}</header>);
};

interface BodyProp {
  nodes: Domain.NodeMarker[];
  cursor: number;
  focused: boolean;
}

/**
 * component definition for body of file list
 */
class Body extends React.Component<BodyProp> {
  render() {
    const {nodes, cursor, focused} = this.props;

    const items = nodes.map((node, index) =>
      (<FileItem key={index} item={node.node} marked={node.marked}
                 selected={index === cursor && focused} />)
    );

    return (
      <ul className="fp-FileList_Content">{items}</ul>
    );
  }
}

interface Prop {
  location: string;
  nodes: Domain.NodeMarker[];
  cursor: number;
  focused: boolean;
}

const FileList : React.FunctionComponent<Prop> = (props) => {
  return (
    <div>
      <Header key="header" directory={props.location} focused={props.focused} />
      <Body key="body" nodes={props.nodes} cursor={props.cursor}
            focused={props.focused} />
    </div>)
};

export default FileList;
