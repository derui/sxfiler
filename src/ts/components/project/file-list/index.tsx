import { h } from "preact";

import * as ListItem from "@/components/project/file-item";
import { AutoSizer } from "@/libs/auto-sizer";

import { ItemMeasureCache } from "./item-measure-cache";
import { ListLayoutCalculator, VirtualizedWindow } from "./list-layout-calculator";
import { FileItem } from "@/generated/filer_pb";
import { Bookmark } from "@/generated/bookmark_pb";
import { useState } from "preact/hooks";

export type Props = {
  location: string;
  items: FileItem.AsObject[];
  bookmarks: Bookmark.AsObject[];
  cursor: number;
  focused: boolean;
};

function useItemMeasureCache() {
  const [cache] = useState(new ItemMeasureCache());
  return cache;
}

const useLayoutCalculator = () => {
  const [calculator] = useState(
    new ListLayoutCalculator({
      estimatedItemSize: 24,
    })
  );

  return calculator;
};

const makeList = (
  height: number,
  props: Props,
  layoutCalculator: ListLayoutCalculator,
  itemMeasureCache: ItemMeasureCache
) => {
  const layout = layoutCalculator.calculateLayout({
    cache: itemMeasureCache,
    currentCursorIndex: props.cursor,
    windowHeight: height,
    listSize: props.items.length,
  });

  return (
    <div class="file-list__list-root" data-testid="fileList-list" style={{ height: height }} key="body">
      {makeListItems(layout, props, itemMeasureCache)}
    </div>
  );
};

const makeListItems = (layout: VirtualizedWindow, props: Props, itemMeasureCache: ItemMeasureCache) => {
  let keyAsPathBookmark = props.bookmarks.reduce((accum, v) => {
    accum[v.path] = v;
    return accum;
  }, {} as { [key: string]: Bookmark.AsObject });

  return props.items.slice(layout.startIndex, layout.stopIndex).map((item, index) => {
    const selected = props.cursor === index + layout.startIndex && props.focused;

    return (
      <ListItem.Component
        key={index + layout.startIndex}
        onRefUpdated={(e: HTMLElement) => itemMeasureCache.set(index + layout.startIndex, e)}
        item={item}
        bookmarked={!!keyAsPathBookmark[item.fullPath]}
        selected={selected}
      />
    );
  });
};

export type HeaderProps = {
  directory: string;
  focused: boolean;
};

/**
 * component definition for header of file list
 */
const Header: preact.FunctionComponent<HeaderProps> = ({ directory, focused }) => {
  return (
    <header class="file-list__header" data-testid="fileList-header" data-focused={focused}>
      {directory}
    </header>
  );
};

export const Component: preact.FunctionComponent<Props> = (props) => {
  const { items, focused, location } = props;
  const cache = useItemMeasureCache();
  const calculator = useLayoutCalculator();

  return (
    <div class="file-list__root" data-testid="fileList-root">
      <Header key="header" directory={location} focused={focused} />
      <AutoSizer class="file-list__autosizer" key="sizer">
        {({ height }) => {
          return items.length === 0 ? (
            <div class="file-list__empty" data-testid="fileList-empty" />
          ) : (
            makeList(height, props, calculator, cache)
          );
        }}
      </AutoSizer>
    </div>
  );
};

export type ElementType = ReturnType<typeof Component>;
