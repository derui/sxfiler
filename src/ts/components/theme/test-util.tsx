import * as React from "react";
import { Theme, ThemeProvider } from "./default-theme";

/**
   wrap element with default theme.
 */
export const wrap = function wrap(e: React.ReactElement) {
  return <ThemeProvider theme={Theme}>{e}</ThemeProvider>;
};
