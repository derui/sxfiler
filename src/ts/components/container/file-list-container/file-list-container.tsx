import * as React from "react";
import { Filer } from "../../../domains/filer";
import { Side, State } from "../../../types/store-state/file-list";
import * as NodeList from "../../project/node-list/node-list";

/* eslint-disable @typescript-eslint/no-var-requires */
const styles: ClassNames = require("./file-list-container.module.scss");
/* eslint-enable @typescript-eslint/no-var-requires */

interface ClassNames {
  root: string;
}

export interface Props {
  state: State;
}

/* create filer from state and key */
function createFiler(key: string, currentSide: Side, filer?: Filer): NodeList.ElementType | null {
  const focused = key === currentSide;

  if (!filer) {
    return null;
  }

  return (
    <NodeList.Component
      key={key}
      nodes={filer.nodes}
      cursor={filer.currentCursorIndex}
      location={filer.location}
      focused={focused}
    />
  );
}

export type ElementType = React.ReactElement<Props, React.FC<Props>>;

// Stateless container to render filer
export const Component: React.FC<Props> = (props): ElementType | null => {
  // can not render anything if filer is not initialized
  if (!props.state.initialized) {
    return null;
  }

  const filers = [
    createFiler(Side.Left, props.state.currentSide, props.state.left),
    createFiler(Side.Right, props.state.currentSide, props.state.right),
  ];

  return <div className={styles.root}>{filers}</div>;
};
