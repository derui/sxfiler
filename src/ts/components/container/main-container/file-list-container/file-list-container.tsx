import * as React from "react";
import { styled } from "@/components/theme";
import { Filer } from "@/domains/filer";
import { Side, State } from "@/states/file-list";
import * as FileList from "@/components/project/file-list";
import * as Element from "@/components/ui/element";

export type Props = {
  state: State;
};

const Root = styled(Element.Component)`
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: 0.5fr auto 0.5fr;

  color: ${props => props.theme.colors.base1};
  overflow: hidden;
  background-color: ${props => props.theme.colors.base03};

  font-size: 1rem;
`;

const Separator = styled.div`
  width: 2px;
  background-color: ${props => props.theme.colors.base00};
  margin: 0 ${props => props.theme.spaces.nano};
`;

/* create filer from state and key */
function createFiler(key: string, currentSide: Side, filer?: Filer): FileList.ElementType | null {
  const focused = key === currentSide;

  if (!filer) {
    return null;
  }

  return (
    <FileList.Component
      key={key}
      items={filer.items}
      cursor={filer.currentCursorIndex}
      location={filer.location}
      focused={focused}
    />
  );
}

export type ElementType = React.ReactElement<Props, React.FC<Props>>;

// Stateless container to render filer
export const Component: React.FC<Props> = (props): ElementType | null => {
  // can not render anything if filer is not initialized
  if (!props.state.left || !props.state.right) {
    return null;
  }

  const leftFiler = createFiler(Side.Left, props.state.currentSide, props.state.left);
  const rightFiler = createFiler(Side.Right, props.state.currentSide, props.state.right);

  return (
    <Root>
      {leftFiler}
      <Separator />
      {rightFiler}
    </Root>
  );
};
