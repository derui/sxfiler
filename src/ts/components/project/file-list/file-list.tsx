import * as React from "react";
import shallowequal from "shallowequal";
import { styled } from "@/components/theme";

import { FileItem } from "@/domains/file-item";
import * as Element from "@/components/ui/element";
import * as List from "@/components/ui/list";
import * as ListItem from "@/components/project/file-item";
import { AutoSizer } from "@/libs/auto-sizer";

import { ItemMeasureCache } from "./item-measure-cache";
import { ListLayoutCalculator, VirtualizedWindow } from "./list-layout-calculator";
import { Component as Header } from "./header";

export type Props = {
  location: string;
  items: FileItem[];
  cursor: number;
  focused: boolean;
};

const Root = styled(Element.Component)`
  display: grid;
  grid-template-rows: calc(${props => props.theme.spaces.base} * 2 + 1rem) 1fr;

  margin: 0;
  padding: 0;

  color: ${props => props.theme.colors.base01};
  overflow: hidden;
  background-color: ${props => props.theme.colors.base03};

  font-size: 1rem;
`;

const AutoSizerWrapper = styled(AutoSizer)`
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
`;

const Empty = styled.div`
  flex: 1 1 auto;
  margin: auto 0;
  text-align: center;

  color: ${props => props.theme.colors.red};

  &:after {
    content: "no any node";
    font-family: monospace;
  }
`;

const ListRoot = styled(List.Component)`
  ${List.style}
  position: relative;

  font-size: 1rem;
  border-right: 1px solid ${props => props.theme.colors.base0};

  &:last-of-type {
    border-right: none;
  }
`;

export type ElementType = React.ReactElement<Props, React.FunctionComponent<Props>>;

export class Component extends React.Component<Props> {
  private itemMeasureCache = new ItemMeasureCache();
  private layoutCalculator: ListLayoutCalculator = new ListLayoutCalculator({
    estimatedItemSize: 24,
  });

  constructor(props: Props) {
    super(props);
  }

  public shouldComponentUpdate(newProps: Props): boolean {
    if (!shallowequal(newProps, this.props)) {
      return true;
    }

    return false;
  }

  private makeList(height: number): JSX.Element {
    const { cursor } = this.props;
    const layout = this.layoutCalculator.calculateLayout({
      cache: this.itemMeasureCache,
      currentCursorIndex: cursor,
      windowHeight: height,
      listSize: this.props.items.length,
    });

    return (
      <ListRoot style={{ height: height }} key="body">
        {this.makeListItems(layout)}
      </ListRoot>
    );
  }

  private makeListItems(layout: VirtualizedWindow): JSX.Element[] {
    const { items, cursor, focused } = this.props;

    return items.slice(layout.startIndex, layout.stopIndex).map((item, index) => {
      const selected = cursor === index + layout.startIndex && focused;
      return (
        <ListItem.Component
          key={index + layout.startIndex}
          ref={e => this.itemMeasureCache.set(index + layout.startIndex, e)}
          item={item}
          selected={selected}
        />
      );
    });
  }

  public render() {
    const { items, focused } = this.props;

    return (
      <Root>
        <Header key="header" directory={this.props.location} focused={focused} />
        <AutoSizerWrapper key="sizer">
          {({ height }) => {
            return items.length === 0 ? <Empty /> : this.makeList(height);
          }}
        </AutoSizerWrapper>
      </Root>
    );
  }
}
