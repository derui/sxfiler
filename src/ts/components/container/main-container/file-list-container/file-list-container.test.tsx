import { mount } from "enzyme";
import * as React from "react";
import renderer from "react-test-renderer";
import { wrap } from "@/components/theme/test-util";

import { Side, State } from "@/states/file-list";
import { Component } from "./file-list-container";
import { createFiler } from "@/domains/filer";
import { createLocationHistory } from "@/domains/location-history";

describe("Container", () => {
  describe("File List Container", () => {
    const history = createLocationHistory({ records: [], maxRecordNumber: 100 });

    it("should render correctly", () => {
      const state: State = {
        left: createFiler({
          id: "left",
          name: "right",
          location: "/left",
          items: [],
          currentCursorIndex: 0,
          history,
        }),
        right: createFiler({
          id: "right",
          name: "right",
          location: "/right",
          items: [],
          currentCursorIndex: 0,
          history,
        }),
        currentSide: Side.Left,
      };
      const tree = renderer.create(wrap(<Component state={state} />)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should mark current side of filer", () => {
      const state: State = {
        left: createFiler({
          id: "left",
          name: "left",
          location: "/left",
          items: [],
          currentCursorIndex: 0,
          history,
        }),
        right: createFiler({
          id: "right",
          name: "right",
          location: "/right",
          items: [],
          currentCursorIndex: 0,
          history,
        }),
        currentSide: Side.Right,
      };
      const wrapper = mount(wrap(<Component state={state} />));

      const focused = wrapper.findWhere(child => child.key() === Side.Right).prop("focused");
      expect(focused).toStrictEqual(true);
    });
  });
});
