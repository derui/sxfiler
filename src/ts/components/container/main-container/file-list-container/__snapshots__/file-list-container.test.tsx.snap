// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container File List Container should render correctly 1`] = `
<div
  className="sc-iwsKbI gGKMtz"
>
  <div
    className="sc-bZQynM gAHUms"
  >
    <header
      className="sc-EHOje eksHn"
      data-focused={true}
    >
      /left
    </header>
    <div
      className="sc-gzVnrw VQBHp"
      style={
        Object {
          "overflow": "visible",
        }
      }
    />
  </div>
  <div
    className="sc-gZMcBi fZsDTv"
  />
  <div
    className="sc-bZQynM gAHUms"
  >
    <header
      className="sc-EHOje eksHn"
      data-focused={false}
    >
      /right
    </header>
    <div
      className="sc-gzVnrw VQBHp"
      style={
        Object {
          "overflow": "visible",
        }
      }
    />
  </div>
</div>
`;
